### PR TITLE
HPCC-24153 add three Machine Learning Docker images build files

### PR DESCRIPTION
**Pull request by xwang2713**
scikit-learning, tensorlow 2, tensorflow 2 with gpu

These images may be used by Intern students for their projects.
I will manually build them per request or fix the broken image, etc

Signed-off-by: Xiaoming Wang <xiaoming.wang@lexisnexis.com>

@jakesmith not sure if you will review the Docker image build. 
And this is not urgent. 

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
All three images are in Docker Hub hpccsystems/platform-core
I did verified libraries for scikit and tf2.

### CODE REVIEW

File: dockerfiles/examples/ml/build.sh [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  'scikit'
+  #'tf2'
+  #'tf2-gpu'
+)
+
+build_image()
+{
+  name=$1
+  ml_label=${name}
+  [[ "${LABEL}" != "latest" ]] && ml_label="${ml_label}-${LABEL}"
```
Comment (jakesmith) [inline dockerfiles/examples/ml/build.sh:50]: confused by handling of "latest". Looks like it will do nothing?
Would clearer if the docker tag block and this block were all inside a != "latest" condition.

File: dockerfiles/examples/ml/scikit/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+# Create base container image to be used by all HPCC processes
+# MORE - some of these dependencies are probably not needed by all derived containers - perhaps we should move them
+# Others may not be wanted at all in container mode - scikit example??
+
+ARG BUILD_LABEL
+FROM hpccsystems/platform-core:${BUILD_LABEL}
+USER root
+
+ENV DEBIAN_FRONTEND=noninteractive
```
Comment (jakesmith) [inline dockerfiles/examples/ml/scikit/Dockerfile:26]: this is in platform-code, so don't think needed again here.

File: dockerfiles/examples/ml/scikit/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+RUN apt clean && \
+    apt autoclean && \
+    apt install -f && \
+    apt autoremove && \
+    apt-get update
```
Comment (jakesmith) [inline dockerfiles/examples/ml/scikit/Dockerfile:31]: is this needed (platform-core already does this).

File: dockerfiles/examples/ml/tf2-gpu/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+ln -s cuda-10.1 /usr/local/cuda && \
+    rm -rf /var/lib/apt/lists/*
+
+# Required for nvidia-docker v1
+RUN echo "/usr/local/nvidia/lib" >> /etc/ld.so.conf.d/nvidia.conf && \
+    echo "/usr/local/nvidia/lib64" >> /etc/ld.so.conf.d/nvidia.conf
+
+
+# Set pip and some required python modules
+ENV DEBIAN_FRONTEND=noninteractive
```
Comment (jakesmith) [inline dockerfiles/examples/ml/tf2-gpu/Dockerfile:53]: don't think this like is needed (as in core)

File: dockerfiles/examples/ml/tf2-gpu/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+ARG CUDNN=7.6.4.38-1
+ARG CUDNN_MAJOR_VERSION=7
+ARG LIB_DIR_PREFIX=x86_64
+ARG LIBNVINFER=6.0.1-1
+ARG LIBNVINFER_MAJOR_VERSION=6
+
+# Needed for string substitution
+SHELL ["/bin/bash", "-c"]
+# Pick up some TF dependencies
+RUN apt-get update && apt-get install -y --no-install-recommends \
```
Comment (jakesmith) [inline dockerfiles/examples/ml/tf2-gpu/Dockerfile:80]: update may need a '-y' (platform-core uses one).

File: dockerfiles/examples/ml/tf2-gpu/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+RUN ln -s /usr/local/cuda/lib64/stubs/libcuda.so /usr/local/cuda/lib64/stubs/libcuda.so.1 \
+    && echo "/usr/local/cuda/lib64/stubs" > /etc/ld.so.conf.d/z-cuda-stubs.conf \
+    && ldconfig
+
+# See http://bugs.python.org/issue19846
+ENV LANG C.UTF-8
+
+#RUN apt-get update && apt-get install -y \
+#    python3 \
+#    python3-pip
```
Comment (jakesmith) [inline dockerfiles/examples/ml/tf2-gpu/Dockerfile:115]: already done, should remove these commented lines.

File: dockerfiles/examples/ml/tf2/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+# Create base container image to be used by all HPCC processes
+# MORE - some of these dependencies are probably not needed by all derived containers - perhaps we should move them
+# Others may not be wanted at all in container mode - tensorflow, etc  ??
+
+ARG BUILD_LABEL
+FROM hpccsystems/platform-core:${BUILD_LABEL}
+USER root
+
+ENV DEBIAN_FRONTEND=noninteractive
```
Comment (jakesmith) [inline dockerfiles/examples/ml/tf2/Dockerfile:26]: as per other comments, don't think this line is necessary

File: dockerfiles/examples/ml/tf2/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+RUN apt clean && \
+    apt autoclean && \
+    apt install -f && \
+    apt autoremove && \
+    apt-get update
```
Comment (jakesmith) [inline dockerfiles/examples/ml/tf2/Dockerfile:31]: don't think this line is necessary

**Review by jakesmith — CHANGES_REQUESTED** [review]

**Comment by jakesmith** [comment]
@xwang2713 - a few comments/questions.

@richardkchapman to can you take a look?

File: dockerfiles/examples/ml/build.sh [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+done
+shift $(( $OPTIND-1 ))
+
+[[ -z ${LABEL} ]] && LABEL=latest
+
+
+ml_features=(
+  'scikit'
+  #'tf2'
+  #'tf2-gpu'
```
[the remark this answers is not in the document]
Comment (xwang2713) [reply dockerfiles/examples/ml/build.sh:43]: Updated as suggested

File: dockerfiles/examples/ml/build.sh (hunk 1 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/build.sh:50]: Here is an example,  let say I want to build tf2. I can pass HPCC Docker Image tag as LEVEL or leave it as blank which will use latest HPCC Docker image.
If use a tag, for example community_7.10.8-1 I will build
1) hpccsystems/platform-core:tf2-community_7.10.8-1
2) if TAGLATEST=1  hpccsystems/platform-core:tf2

if use latest HPCC Docker image (LABEL=latest)
only  hpccsystems/platform-core:tf2 will be created.

File: dockerfiles/examples/ml/scikit/Dockerfile (hunk 2 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/scikit/Dockerfile:26]: removed

File: dockerfiles/examples/ml/scikit/Dockerfile (hunk 3 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/scikit/Dockerfile:31]: Just in chance base image apt install need clean-up.
Prefer to keep them. But can be removed also.

File: dockerfiles/examples/ml/tf2-gpu/Dockerfile (hunk 4 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/tf2-gpu/Dockerfile:53]: done

File: dockerfiles/examples/ml/tf2-gpu/Dockerfile (hunk 5 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/tf2-gpu/Dockerfile:80]: done

Comment (xwang2713) [reply dockerfiles/examples/ml/tf2-gpu/Dockerfile:80]: Add "-y" to other apt update also

File: dockerfiles/examples/ml/tf2-gpu/Dockerfile (hunk 6 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/tf2-gpu/Dockerfile:115]: done

File: dockerfiles/examples/ml/tf2/Dockerfile (hunk 7 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/tf2/Dockerfile:26]: done

File: dockerfiles/examples/ml/tf2/Dockerfile (hunk 8 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/tf2/Dockerfile:31]: As mentioned before prefer to keep them. But can be removed.

**Comment by xwang2713** [comment]
Per a intern student request I add scikit-learn to tf2 and tf2-gpu since it is useful and small compared with tensorflow packages.
I will test above changes and update the code of this PR.

**Comment by xwang2713** [comment]
@jakesmith updated PR. See my comment for your each of your review.

File: dockerfiles/examples/ml/scikit/Dockerfile (hunk 3 above)
Comment (jakesmith) [reply dockerfiles/examples/ml/scikit/Dockerfile]: >Just in chance base image apt install need clean-up.

by definition it's clean I think, so don't know why it would need autoclean , install -f , autoremove, update.

I think should remove, unless reason is clear why needed.

File: dockerfiles/examples/ml/build.sh [hunk 10]
```diff
@@ -19,30 +19,54 @@
...
+  'tf2-gpu'
 )
 
+found="false"
+for ml_feature in ${ml_features[@]}
+do
+	if [[ $ml_feature ==  $FEATURE ]]
+	then
+          found="true"
+	  break
```
Comment (jakesmith) [inline dockerfiles/examples/ml/build.sh:57]: indentation a little odd, I think there are tabs here (that should be replaced by spaces)

File: dockerfiles/examples/ml/tf2/Dockerfile (hunk 8 above)
Comment (jakesmith) [reply dockerfiles/examples/ml/tf2/Dockerfile]: see prev. reply, think should remove.

**Comment by jakesmith** [comment]
@xwang2713 - apart from a couple of v. minor comments, I think it looks fine.

But would like @richardkchapman to take a look also.

File: dockerfiles/examples/ml/build.sh (hunk 10 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/build.sh:57]: fixed

**Comment by xwang2713** [comment]
I remove "apt autoclean  && ... apt update" except the one in tf2-gpu/Dockerfiles. Otherwise "python3-pip" can be founded and installed.

Tested with 7.10.10-1

File: dockerfiles/examples/ml/build.sh [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+found="false"
+for ml_feature in ${ml_features[@]}
+do
+	if [[ $ml_feature ==  $FEATURE ]]
+	then
+    found="true"
+	  break
+	fi
+done
+if [[ "$found" == "flase" ]]
```
Comment (richardkchapman) [inline dockerfiles/examples/ml/build.sh:60]: Is this supposed to be "false" ?

**Comment by richardkchapman** [comment]
Should we be using the label to indicate what has been added, or should they be different image names?

i.e. would it make more sense to be building platform-tf2:7.10.10 rather than platform-core:tf2-7.10.10 ?

The build.sh should act like buildall.sh in dockerfile directory (or perhaps we should be doing these builds inside buildall) so that the images are built and pushed automatically via github actions when a new build is tagged

File: dockerfiles/examples/ml/scikit/Dockerfile (hunk 3 above)
Comment (richardkchapman) [reply dockerfiles/examples/ml/scikit/Dockerfile]: You don't need them if you are in lock-step with the core - i.e. this is built every time the core is built.

But if you are building something based on a core built some time back, you might.

File: dockerfiles/examples/ml/build.sh (hunk 11 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/build.sh]: fixed

**Comment by xwang2713** [comment]
For Docker Image name maybe (ml features):
1) platform-ml:   general machine learning image. Currently include scikit-learn
2) platform-tf2:   tensorflow + scikit-learn
3) platform-tf2-gpu:  tensorflow + gpu + scikit-learn

I will add ml build to buildall.sh and add an option to enable it (default is disable, I am afraid it takes too long to build). Maybe a n option for the list of ml features.

**Comment by xwang2713** [comment]
To build ML in buildall.sh set environment variable BUILD_ML=all

**Comment by richardkchapman** [comment]
@xwang2713 Please squash. Looks ready apart from that

**Comment by xwang2713** [comment]
@richardkchapman,  done.

File: dockerfiles/examples/ml/README.md [hunk 12]
```diff
@@ -0,0 +1,53 @@
+# HPCC Systems Machine Learning Docker Images
```
Comment (richardkchapman) [inline dockerfiles/examples/ml/README.md:1]: Now that we are building these from buildall, I'm not sure that 'examples' is the right place for the files...

**Comment by richardkchapman** [comment]
@xwang2713 One last thing...

File: dockerfiles/examples/ml/README.md (hunk 12 above)
Comment (xwang2713) [reply dockerfiles/examples/ml/README.md]: Move ml from dockerfile/examples/ to dockerfiles/ and update the ml path inside buildall.sh script.
